In [8]:
import pandas as pd 
import re # re → text cleaning (very important now)
df = pd.read_csv("/kaggle/input/datasets/lakshmi25npathi/imdb-dataset-of-50k-movie-reviews/IMDB Dataset.csv")
df.head()
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   review     50000 non-null  object
 1   sentiment  50000 non-null  object
dtypes: object(2)
memory usage: 781.4+ KB


**Text Cleaning (VERY IMPORTANT UPGRADE)**

In [9]:
def clean_text(text):
    text = text.lower() #"Amazing" and "amazing" are treated as different words
    #That’s bad because:
    #same meaning, but model thinks different
    text = re.sub(r"[^a-zA-Z]", " ", text)
    return text

df["review"] = df["review"].apply(clean_text)

**2. re.sub(r"[^a-zA-Z]", " ", text)

This is the main cleaning step
🧠 Understand the regex:
[^a-zA-Z]

Means:
👉 “Remove everything that is NOT a letter”

So it removes:

Numbers → 123
Symbols → ! @ # %
Punctuation → . , ?
👉 Example:

Before:

"Movie was AMAZING!!! 10/10 😍"

After:
"movie was amazing"**

🔸 3. .apply(clean_text)
df["review"] = df["review"].apply(clean_text)


👉 This applies your function to every row in the column

So:

Row 1 → cleaned
Row 2 → cleaned
Row 3 → cleaned
…

**Feature Extraction (TF-IDF — KEY STEP)**

In [10]:
from sklearn.feature_extraction.text import TfidfVectorizer

tfidf = TfidfVectorizer(max_features=5000)
X = tfidf.fit_transform(df["review"])
y = df["sentiment"]

🔥 TF-IDF makes it smarter

Instead of just counting…

👉 It gives importance (weight) to words

🧠 TF-IDF = Two things
🔹 1. TF (Term Frequency)

👉 How often word appears in a sentence

🔹 2. IDF (Inverse Document Frequency)

👉 How rare the word is across all sentences

🔥 Key Idea:
Word	Meaning	Importance
"the"	common	❌ low
"movie"	common	❌ low
"amazing"	rare	✅ high
"terrible"	rare	✅ high
🔸 Final Result

Now your vector looks like:

"movie is good" → [0.2, 0.1, 0.9, 0.0]


👉 “good” gets higher value
👉 “is”, “movie” get lower value

➡️ Model focuses on important words        


tfidf = TfidfVectorizer(max_features=5000)
👉 max_features=5000
Keeps top 5000 important words only

Removes less useful words

👉 Why?

Reduces memory

Avoids noise

Speeds up training

**Train-Test Split**

In [11]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42
)

**Model Training**

In [12]:
from sklearn.linear_model import LogisticRegression

model = LogisticRegression()
model.fit(X_train, y_train)

LogisticRegression()

**Prediction + Accuracy**

In [13]:
from sklearn.metrics import accuracy_score

y_pred = model.predict(X_test)

accuracy = accuracy_score(y_test, y_pred)
print("Accuracy:", accuracy)

Accuracy: 0.8945


**Compare with Old Method**

**CountVectorizer Accuracy: 84%**

**TF-IDF Accuracy: 89%**


## Improvement

- TF-IDF improved performance compared to CountVectorizer
- Cleaning text improved model quality

## Conclusion

Model can effectively classify sentiment from text data
